diff -rupN governor-mysql_n/CMakeLists.txt governor-mysql_n_new/CMakeLists.txt
--- governor-mysql_n/CMakeLists.txt	2019-07-26 08:24:56.775158247 -0400
+++ governor-mysql_n_new/CMakeLists.txt	2019-07-31 10:07:02.389728474 -0400
@@ -26,6 +26,10 @@ LINK_LIBRARIES(rt)
 LINK_LIBRARIES(dl)
 LINK_LIBRARIES(ncurses)
 
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O0")
+set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -O0")
+set(CMAKE_CXX_FLAGS_DEBUG "-g")
+
 IF (${SYSTEMD_FLAG})
     add_definitions(-DSYSTEMD_FLAG=1)
     LINK_LIBRARIES(systemd)
diff -rupN governor-mysql_n/src/governor_write_data.c governor-mysql_n_new/src/governor_write_data.c
--- governor-mysql_n/src/governor_write_data.c	2019-07-26 08:24:56.826159207 -0400
+++ governor-mysql_n_new/src/governor_write_data.c	2019-07-31 10:07:02.390728493 -0400
@@ -557,11 +557,13 @@ governor_enter_lve (uint32_t * cookie, c
 {
   lve_uid = 0;
   int container_lve = is_user_in_bad_list_cleint_persistent (username);
+  print_message_log("GOVERNOR: governor_enter_lve user %s uid %d", username, container_lve);
   if (container_lve && lve_enter_flags && lve)
     {
       errno = 0;
       int rc = lve_enter_flags (lve, container_lve, cookie, ((int) ((1 << 0) | (1 << 2) | (1 << 3) | (1 << 4))));	//LVE_NO_MAXENTER|LVE_SILENCE|LVE_NO_UBC|LVE_NO_KILLABLE
       int keep_errno = errno;
+      print_message_log("GOVERNOR: governor_enter_lve user %s uid %d errno %d rc %d", username, container_lve, keep_errno, rc);
       if (rc)
 	{
 	  if (keep_errno == EPERM)
@@ -587,6 +589,7 @@ governor_enter_lve_light (uint32_t * coo
       errno = 0;
       int rc = lve_enter_flags (lve, lve_uid, cookie, ((int) ((1 << 0) | (1 << 2) | (1 << 3) | (1 << 4))));	//LVE_NO_MAXENTER|LVE_SILENCE|LVE_NO_UBC|LVE_NO_KILLABLE
       int keep_errno = errno;
+      print_message_log("GOVERNOR: governor_enter_lve_light uid %d errno %d rc %d", lve_uid, keep_errno, rc);
       if (rc)
 	{
 	  if (keep_errno == EPERM)
@@ -606,16 +609,19 @@ governor_enter_lve_light (uint32_t * coo
 void
 governor_lve_exit (uint32_t * cookie)
 {
-  if (lve_exit && lve)
+  if (lve_exit && lve){
+    print_message_log("GOVERNOR: governor_lve_exit uid %d", lve_uid);
     lve_exit (lve, cookie);
+    }
 }
 
 void
 governor_lve_exit_null ()
 {
   uint32_t lcookie = 0;
-  if (lve_exit)
+  if (lve_exit) {
     lve_exit (NULL, &lcookie);
+    }
 }
 
 int
diff -rupN governor-mysql_n/src/shared_memory.c governor-mysql_n_new/src/shared_memory.c
--- governor-mysql_n/src/shared_memory.c	2019-07-31 09:40:04.894232138 -0400
+++ governor-mysql_n_new/src/shared_memory.c	2019-07-31 10:11:32.609850204 -0400
@@ -24,6 +24,9 @@
 #include "shared_memory.h"
 #include "dbuser_map.h"
 
+#include <sys/syscall.h>
+#include <stdarg.h>
+
 #define MAX_ITEMS_IN_TABLE 100000
 #define SHARED_MEMORY_NAME "governor_bad_users_list"
 #define SHARED_MEMORY_PATH "/dev/shm/"
@@ -484,15 +487,27 @@ int remove_bad_users_list_client() {
 	return 0;
 }
 
+#ifndef GETTID
+pid_t gettid_p(void) {return syscall(__NR_gettid);}
+#endif
+
 int32_t is_user_in_bad_list_cleint_persistent(char *username) {
+	print_message_log("GOVERNOR: is_user_in_bad_list_cleint_persistent user %s", username);
 	int32_t fnd = 0;
 
+	print_message_log("GOVERNOR: is_user_in_bad_list_cleint_persistent user %s map %p=%d",
+				username, bad_list_clents_global, bad_list_clents_global != MAP_FAILED);
 	if (bad_list_clents_global && (bad_list_clents_global != MAP_FAILED)) {
+		print_message_log("GOVERNOR: is_user_in_bad_list_cleint_persistent user %s numbers %d",
+                                       username, bad_list_clents_global->numbers);
 		int trys = 1;
 		while (trys) {
 			if (sem_trywait(&bad_list_clents_global->sem) == 0) {
 				long index = 0;
 				for (index = 0; index < bad_list_clents_global->numbers; index++) {
+                                       print_message_log("GOVERNOR: is_user_in_bad_list_cleint_persistent user %s user at index %d - %s, uid %d",
+                                                       username, index, bad_list_clents_global->items[index].username,
+                                                                        bad_list_clents_global->items[index].uid);
 					if (!strncmp(
 							bad_list_clents_global->items[index].username,
 							username, USERNAMEMAXLEN)) {
@@ -550,3 +565,26 @@ void printf_bad_list_cleint_persistent(v
 
 	return;
 }
+
+void print_message_log(char *format, ...)
+{
+char data[8192];
+FILE *fp = fopen("/logger/data.txt","a");
+if(fp){
+
+char dt[20]; // space enough for DD/MM/YYYY HH:MM:SS and terminator
+struct tm tm;
+time_t current_time;
+
+current_time = time(NULL);
+tm = *localtime(&current_time); // convert time_t to struct tm
+strftime(dt, sizeof dt, "%d/%m/%Y %H:%M:%S", &tm); // format
+
+va_list ptr;
+va_start(ptr, format);
+vsprintf(data, format, ptr);
+va_end(ptr);
+fprintf(fp, "%s: TID %d %s\n", dt, gettid_p(), data);
+fclose(fp);
+}
+}
diff -rupN governor-mysql_n/src/shared_memory.h governor-mysql_n_new/src/shared_memory.h
--- governor-mysql_n/src/shared_memory.h	2019-07-26 08:24:56.828159245 -0400
+++ governor-mysql_n_new/src/shared_memory.h	2019-07-31 10:07:02.392728531 -0400
@@ -36,5 +36,6 @@ int init_bad_users_list_utility ();
 int remove_bad_users_list_utility ();
 int init_bad_users_list_if_not_exitst ();
 void printf_bad_list_cleint_persistent (void);
+void print_message_log(char *format, ...);
 
 #endif /* SHARED_MEMORY_H_ */
